tools/fp20compiler/main.cpp

```diff
@@ -34,30 +34,124 @@ nvparse_errors errors;
-    if (ts_start) {
-        char* ts_end = strstr(ts_start, "// End of program");
-        assert(ts_end);
-        *ts_end = 0;
+        // Accept if candidate is at beginning of file or line
+        if ((cursor == haystack) || (cursor[-1] == '\n')) {
+            return (char*)cursor;
+        }
 
-        ts10_init(ts_start);
-        ts10_parse();
+        cursor = strstr(cursor + 1, needle);
+    }
+    return NULL;
+}
 
-        *ts_end = '/';
+static char* skip_until(const char* str, const char* delim) {
+    while(*str != '\0') {
+        const char* symbol = delim;
+        while(*symbol != '\0') {
+            if (*str == *symbol++) {
+                return (char*)str;
+            }
+        }
+        str++;
     }
+    return (char*)str;
+}
 
-    char* rc_start = strstr(ns, "!!RC1.0");
-    if (rc_start) {
-        rc10_init(rc_start);
-        rc10_parse();
+void translate(const char* s) {
+
+    // Keep a cursor for line-counting
+    const char* line_cursor = s;
+    unsigned int line = 1;
+
+    // Look for the first shader magic
+    const char* shader_magic = find_at_line_start(s, s, "!!");
+
+    // Warn the user if we couldn't find any shader at all
+    if (shader_magic == NULL) {
+        fprintf(stderr, "no shaders found\n");
     }
 
-    free(ns);
+    // Loop until we can't find a shader anymore
+    while (shader_magic != NULL) {
+
+        // Find end of shader magic line (whitespace or comment)
+        const char* shader_magic_end = shader_magic+2;
+        while(1) {
+            shader_magic_end = skip_until(shader_magic_end, " \t\n\r/#");
+            if (shader_magic_end[0] == '/') {
+                if ((shader_magic_end[1] != '*') &&
+                    (shader_magic_end[1] != '/')) {
+                    shader_magic_end++;
+                    continue;
+                }
+            }
+            break;
+        }
+
+        // Copy the magic
+        size_t shader_magic_len = shader_magic_end - shader_magic;
+        char* shader_magic_str = copy_string(shader_magic, shader_magic_len);
+
+        // Count which line the current shader starts on
+        while (line_cursor < shader_magic) {
+            line_cursor = strchr(line_cursor, '\n');
+            assert(line_cursor != NULL);
+            line_cursor++;
+            line++;
+        }
+
+        // Add information about shader section to output
+        printf("/* %s (line %u) */\n", shader_magic_str, line);
+
+        // Mark shader magic as shader start
+        const char* shader = shader_magic;
+
+        // The next shader magic marks the end of current shader; find it
+        const char* next_shader_magic;
+        const char* shader_end;
+        if (*shader_magic_end != '\0') {
+            next_shader_magic = find_at_line_start(s, shader_magic_end+1, "!!");
+            shader_end = next_shader_magic;
+        } else {
+            next_shader_magic = NULL;
+            shader_end = shader_magic_end;
```

Comment:
This looks bad?

This must probably be `&shader_magic_end[1+strlen(shader_magic_end-1)];`
Basically search for end of file, starting where the magic ends.